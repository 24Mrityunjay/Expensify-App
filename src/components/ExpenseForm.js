import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/initialize';

export default class ExpenseForm extends React.Component{
    constructor(props){
        super(props);
        this. state = {
            description: props.expense ? props.expense.description : "",
            note: props.expense ? props.expense.note : "",
            createdAt:props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused:false,
            error:'',
            amount: props.expense ? (props.expense.amount).toString() : ""
        };
    }
      
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(()=>({description}));
    };
    onTextAreaChange = (e) => {
        e.persist();
        this.setState(()=>({note : e.target.value}))
    };
    onAmountChange = (e) => {
        const amount = e.target.value;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(()=>({amount}));
        }
    };
    onDateChange = (createdAt) => {
        this.setState(() => ({ createdAt }));
    };
    onFocusChange = ({focused}) => {
        this.setState(()=>({calendarFocused: focused}));
    };
    onSubmit = (e) => {
        e.preventDefault();
        if(!this.state.description || !this.state.amount){
            this.setState(()=>({
                error:"Please provide description and amount"
            }));
            //alert("Please provide description and amount");
        }else{
            this.setState(()=>({
                error:""
            }));
            this.props.onSubmit({
                description:this.state.description,
                amount:parseFloat(this.state.amount, 10) ,
                note:this.state.note,
                createdAt:this.state.createdAt.valueOf()
            });
            //remove error
            console.log("Form Submitted Successfully!");
            // this.setState(()=>({
            //     description:"",
            //     note:'',
            //     createdAt:moment(),
            //     calendarFocused:false,
            //     error:'',
            //     amount:''
            // }));
        }
    };
    render(){
        return (
            <div>
                <h1>ExpenseForm</h1>
                {this.state.error && <span style={{color: "red"}}>{this.state.error}</span>}
                <form onSubmit={this.onSubmit}>
                 <div>   
                     <input 
                        type="text"
                        placeholder="AddExpense"
                        value={this.state.description}
                        onChange={this.onDescriptionChange}                        
                        autoFocus
                     />
                     <input 
                        type="number"
                        placeholder="Enter Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                     /><br/>
                     
                 </div>
                 <div>
                     <textarea
                        placeholder="Enter a note for your Expense(optional)"
                        value={this.state.note}
                        onChange={this.onTextAreaChange}
                     ></textarea>
                 </div>
                 <div>
                     <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths= {1}
                        isOutsideRange={()=>false}
                     /><br/>   
                       
                 </div>          
                    <button>Add Expense</button>
                </form>  
            </div>
        )
    }
}