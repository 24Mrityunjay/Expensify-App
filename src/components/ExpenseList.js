import React from 'react';
import { connect } from 'react-redux';  
import ExpenseListItem  from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';
import ExpenseListFilters from './ExpenseListFilters';

export const ExpenseList = (props) =>(
    <div>
        <h1>Expense List</h1>
        {
            props.expenses.length === 0 ? <h1>No expenses</h1> : (
                props.expenses.map((expense)=>{
                    return <ExpenseListItem key={expense.id} {...expense}/>;
                })
            )
        }

        <ExpenseListFilters/>
    
    </div>
);

const mapsToFunc = ((state)=> {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
});

export default connect(mapsToFunc)(ExpenseList);
