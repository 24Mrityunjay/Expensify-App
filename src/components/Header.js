import React from 'react';
import { NavLink} from 'react-router-dom';

const Header = ()=>(
    <header>
       <h1> Expense App</h1> 
       <NavLink to='/' activeClassName='is-active' exact={true }>Go Home</NavLink><br/>
       <NavLink to='/create' activeClassName='is-active'>Go To Create Page</NavLink><br/>
       <NavLink to='/edit' activeClassName='is-active'>Go To Edit Page</NavLink><br/>  
    </header>
);

export default Header;