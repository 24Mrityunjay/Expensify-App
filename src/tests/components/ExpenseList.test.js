import React from 'react';
import { shallow } from 'enzyme';
import ReactShallowRenderer  from "react-test-renderer/shallow";
import { ExpenseList } from '../../components/ExpenseList';
//import expenses from '../fixtures/expenses';

export const expenses = [{
    id: '1',
    description: 'Gum',
    note: '', 
    amount: 195,
    createdAt: 20
}, {
    id: '2',
    description: 'Rent',
    note: '', 
    amount: 109500,
    createdAt: 1000
}, {
    id: '3',
    description: 'Credit Card',
    note: '', 
    amount: 4500,
    createdAt: -1000
}];

test('Should refer expensesList with expenses', () => {
    // const wrapper = shallow(<ExpenseList expenses={expenses}/>);
    // expect(wrapper).toMatchSnapshot();

    const renderer = new ReactShallowRenderer();
    renderer.render(<ExpenseList expenses={expenses}/>);
    expect(renderer.getRenderOutput()).toMatchSnapshot();
});

test('Should refer expensesList with expenses', () => {
    // const wrapper = shallow(<ExpenseList expenses={[]}/>);
    // expect(wrapper).toMatchSnapshot();

    const renderer = new ReactShallowRenderer();
    renderer.render(<ExpenseList expenses={[]}/>);
    expect(renderer.getRenderOutput()).toMatchSnapshot();
});

