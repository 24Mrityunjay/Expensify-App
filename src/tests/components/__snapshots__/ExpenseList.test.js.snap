// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should refer expensesList with expenses 1`] = `
<div>
  <h1>
    Expense List
  </h1>
  <ExpenseListItem
    amount={195}
    createdAt={20}
    description="Gum"
    id="1"
    note=""
  />
  <ExpenseListItem
    amount={109500}
    createdAt={1000}
    description="Rent"
    id="2"
    note=""
  />
  <ExpenseListItem
    amount={4500}
    createdAt={-1000}
    description="Credit Card"
    id="3"
    note=""
  />
  <Memo(Connect(ExpenseListFilters)) />
</div>
`;

exports[`Should refer expensesList with expenses 2`] = `
<div>
  <h1>
    Expense List
  </h1>
  <h1>
    No expenses
  </h1>
  <Memo(Connect(ExpenseListFilters)) />
</div>
`;
